### PR TITLE
[Feature] Pointwise arithmetic operations using foreach

### PR DESCRIPTION
**Pull request by vmoens**
Test the feature:
```python
from tensordict import TensorDict
from torch.nn import Transformer
import torch
from torch.utils.benchmark import Timer
t = Transformer(device="cuda" if torch.cuda.is_available() else "cpu")

params = TensorDict.from_module(t).apply(lambda t: t.data.clone().fill_(0.0))
other_params = TensorDict.from_module(t).apply(lambda t: t.data.clone().fill_(1.0))

assert (params.add(other_params) == 1).all()

print("\n\nadd")
params.add(other_params)
print("foreach", Timer("params.add(other_params)", globals=globals()).adaptive_autorange())
params.lock_()
other_params.lock_()
params.add(other_params)
print("foreach, locked", Timer("params.add(other_params)", globals=globals()).adaptive_autorange())
params.apply(lambda x, y: x+y, other_params)
print("apply", Timer("params.apply(lambda x, y: x+y, other_params)", globals=globals()).adaptive_autorange())
listval0 = params._values_list(True, True)
listval1 = other_params._values_list(True, True)
torch._foreach_add(listval0, listval1)
print("plain foreach", Timer("torch._foreach_add(listval0, listval1)", globals=globals()).adaptive_autorange())

params.unlock_()
other_params.unlock_()

print("\n\nadd_")
params.add(other_params)
print("foreach", Timer("params.add_(other_params)", globals=globals()).adaptive_autorange())
params.lock_()
other_params.lock_()
params.add(other_params)
print("foreach, locked", Timer("params.add_(other_params)", globals=globals()).adaptive_autorange())
params.apply(lambda x, y: x+y, other_params)
print("apply", Timer("params.apply(lambda x, y: x.add_(y), other_params)", globals=globals()).adaptive_autorange())
listval0 = params._values_list(True, True)
listval1 = other_params._values_list(True, True)
torch._foreach_add_(listval0, listval1)
print("plain foreach", Timer("torch._foreach_add_(listval0, listval1)", globals=globals()).adaptive_autorange())
```

Gives these results:

```
add
foreach <torch.utils.benchmark.utils.common.Measurement object at 0x7f51c7f5ab60>
params.add(other_params)
  Median: 3.62 ms
  IQR:    0.07 ms (3.59 to 3.66)
  4 measurements, 100 runs per measurement, 1 thread
foreach, locked <torch.utils.benchmark.utils.common.Measurement object at 0x7f5169b13e20>
params.add(other_params)
  Median: 1.94 ms
  IQR:    0.12 ms (1.93 to 2.05)
  4 measurements, 100 runs per measurement, 1 thread
apply <torch.utils.benchmark.utils.common.Measurement object at 0x7f5169b11840>
params.apply(lambda x, y: x+y, other_params)
  Median: 4.18 ms
  IQR:    0.02 ms (4.17 to 4.19)
  4 measurements, 100 runs per measurement, 1 thread
plain foreach <torch.utils.benchmark.utils.common.Measurement object at 0x7f5169b12d40>
torch._foreach_add(listval0, listval1)
  Median: 567.85 us
  IQR:    8.00 us (563.76 to 571.76)
  4 measurements, 100 runs per measurement, 1 thread


add_
foreach <torch.utils.benchmark.utils.common.Measurement object at 0x7f51c7f59cc0>
params.add_(other_params)
  Median: 929.06 us
  IQR:    9.48 us (925.40 to 934.88)
  4 measurements, 100 runs per measurement, 1 thread
foreach, locked <torch.utils.benchmark.utils.common.Measurement object at 0x7f5169b11540>
params.add_(other_params)
  Median: 353.06 us
  IQR:    0.10 us (353.00 to 353.10)
  4 measurements, 1000 runs per measurement, 1 thread
apply <torch.utils.benchmark.utils.common.Measurement object at 0x7f5169b13f10>
params.apply(lambda x, y: x.add_(y), other_params)
  Median: 3.34 ms
  IQR:    0.02 ms (3.32 to 3.34)
  4 measurements, 100 runs per measurement, 1 thread
plain foreach <torch.utils.benchmark.utils.common.Measurement object at 0x7f5169b107f0>
torch._foreach_add_(listval0, listval1)
  Median: 353.14 us
  IQR:    0.07 us (353.10 to 353.17)
  4 measurements, 1000 runs per measurement, 1 thread
```

So there is some overhead for add due to the fact that we need to rebuild the tensordict, but if we don't (add_) it's as fast as plain foreach and you can keep the nice structure of your params

cc @fmassa @ahmed-touati @teopir @MateuszGuzek @zou3519 @ezyang @albanD

### CODE REVIEW

**Comment by vmoens** [comment]
@janeyx99 @fmassa 

Here is a version of Adam in 23 lines of code (no weight decay and no edge case whatsoever) with tensordict.
With this you can easily recycle your adam for a sub-sub module, just index the weights accordingly, e.g. `self.weight["decoder", "0"]` 

```python
import torch

from tensordict import TensorDict
from tensordict.nn import TensorDictParams
from torch import nn

class Adam:
    def __init__(self, weights: TensorDictParams, alpha: float=1e-3, beta1: float=0.9, beta2: float=0.999, eps: float = 1e-6, weight_decay: float=0.0):
        # Lock for efficiency
        weights = weights.lock_()
        self.weights = weights
        self.t = 0

        self._mu = weights.data.clone()
        self._sigma = weights.data.mul(0.0)
        self.beta1 = beta1
        self.beta2 = beta2
        self.alpha = alpha
        self.eps = eps

    def step(self):
        self._mu.mul_(self.beta1).add_(self.weights.grad, 1 - self.beta1)
        self._sigma.mul_(self.beta2).add_(self.weights.grad.pow(2), 1 - self.beta2)
        self.t += 1
        mu = self._mu.div_(1-self.beta1**self.t)
        sigma = self._sigma.div_(1 - self.beta2 ** self.t)
        self.weights.data.add_(mu.div_(sigma.sqrt_().add_(self.eps)).mul_(-self.alpha))

    def zero_grad(self):
        self.weights.grad.zero_()

device = "cpu" if not torch.cuda.is_available() else "cuda"
with torch.device(device):
    net = nn.Transformer()
    weights = TensorDict.from_module(net, as_module=True)
    adam = Adam(weights=weights)
    net(torch.randn(5, 5, 512), torch.randn(5, 5, 512)).mean().backward()
    adam.step()
    adam.zero_grad()

    from torch.utils.benchmark import Timer
    net(torch.randn(5, 5, 512), torch.randn(5, 5, 512)).mean().backward()
    print("td", Timer("adam.step()", globals=globals()).adaptive_autorange())
    adam.zero_grad()

    adam_orig = torch.optim.Adam(params=net.parameters())

    net(torch.randn(5, 5, 512), torch.randn(5, 5, 512)).mean().backward()
    adam_orig.step()
    adam_orig.zero_grad()

    net(torch.randn(5, 5, 512), torch.randn(5, 5, 512)).mean().backward()
    print("native", Timer("adam.step()", globals=globals()).adaptive_autorange())
    adam.zero_grad()
```

On my machine and on A100 I get a similar runtime (again, this is much simpler than torch.optim - my point is that at least it isn't slower!)

**Comment by vmoens** [comment]
@matteobettini 

With this you can do
```
td += 1
```
with a lazy stack and it's efficient (much more efficient than stack -> do the op -> call __setitem__ which was the default behaviour before). That will allow us to do ops on lazy stacks as if they were regular contiguous stacks of tensors

**Comment by janeyx99** [comment]
Just to make sure I understand your point--this is showing that foreach is built into how TensorDicts can get one thing applied to multiple params at once?
